### PR TITLE
Add support for federation on browser

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Add support for federation on the browser

## Motivation and Context
Add support for federation on the browser

## Regression
Yes not sure

## How Has This Been Tested?
By setting up the federation setup using the federation docs https://docs.minio.io/docs/minio-federation-quickstart-guide.html

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

File: cmd/generic-handlers.go [hunk 1]
```diff
@@ -631,6 +631,46 @@ func (f bucketForwardingHandler) ServeHTTP(w http.ResponseWriter, r *http.Reques
 		f.handler.ServeHTTP(w, r)
 		return
```
Comment (nitisht) [inline cmd/generic-handlers.go:632]: Shouldn't we remove the `guessIsBrowserReq` check?

Comment (harshavardhana) [reply cmd/generic-handlers.go:632]: yes, correct - it was brought in after rebase.

**Comment by harshavardhana** [comment]
The pending item in this PR is still RemoveObjects() hopefully will be done by tomorrow

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :x: [more...](#6b6495a1-37c6-4a6c-a880-a180cb46ece3) |
<a name="6b6495a1-37c6-4a6c-a880-a180cb46ece3"></a>
### 6891-10eec50/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31342
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c823c1c2d5fc:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 51 seconds
(2/13) Running aws-sdk-go tests ... done in 1 seconds
(3/13) Running aws-sdk-java tests ... done in 5 seconds
(4/13) Running aws-sdk-php tests ... done in 41 seconds
(5/13) Running aws-sdk-ruby tests ... done in 8 seconds
(6/13) Running mc tests ... done in 44 seconds
(7/13) Running minio-dotnet tests ... done in 2 minutes and 49 seconds
(8/13) Running minio-go tests ... done in 48 seconds
(9/13) Running minio-java tests ... done in 2 minutes and 58 seconds
(10/13) Running minio-js tests ... FAILED in 36 seconds
{
  "name": "minio-js",
  "function": "listenBucketNotification(bucketName, prefix, suffix, events)",
  "args": "bucketName:minio-js-test-d7e7cf93-48d4-438c-9b4a-aebae76b31bc, events: s3:ObjectCreated:*",
  "duration": 20100,
  "status": "FAIL",
  "error": "Error: notification not received at client.removeObject.err (test/functional-tests.js:1100:43) at ClientRequest.<anonymous> (node_modules/minio/dist/main/minio.js:419:9) at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:478:21) at HTTPParser.parserOnHeadersComplete (_http_common.js:99:23) at Socket.socketOnData (_http_client.js:367:20) at readableAddChunk (_stream_readable.js:176:18) at Socket.Readable.push (_stream_readable.js:134:10) at TCP.onread (net.js:559:20)"
}

Executed 9 out of 13 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :x: [more...](#c1726c9e-dd26-4d10-9d77-70ae408897be) |
| mint-dist-xl.sh | :x: [more...](#67498928-0513-469c-a122-ad1c6984036c) |
<a name="c1726c9e-dd26-4d10-9d77-70ae408897be"></a>
### 6891-b8e7928/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30148
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 7b0103e282d0:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 3 minutes and 49 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 7 seconds
(4/13) Running aws-sdk-php tests ... done in 42 seconds
(5/13) Running aws-sdk-ruby tests ... done in 2 seconds
(6/13) Running mc tests ... done in 28 seconds
(7/13) Running minio-dotnet tests ... done in 2 minutes and 50 seconds
(8/13) Running minio-go tests ... done in 34 seconds
(9/13) Running minio-java tests ... done in 3 minutes and 18 seconds
(10/13) Running minio-js tests ... FAILED in 29 seconds
{
  "name": "minio-js",
  "function": "listenBucketNotification(bucketName, prefix, suffix, events)",
  "args": "bucketName:minio-js-test-e39b268c-094c-4fd0-a509-418462f27312, events: s3:ObjectCreated:*",
  "duration": 20099,
  "status": "FAIL",
  "error": "Error: notification not received at client.removeObject.err (test/functional-tests.js:1100:43) at ClientRequest.<anonymous> (node_modules/minio/dist/main/minio.js:419:9) at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:478:21) at HTTPParser.parserOnHeadersComplete (_http_common.js:99:23) at Socket.socketOnData (_http_client.js:367:20) at readableAddChunk (_stream_readable.js:176:18) at Socket.Readable.push (_stream_readable.js:134:10) at TCP.onread (net.js:559:20)"
}

Executed 9 out of 13 tests successfully.
```
<a name="67498928-0513-469c-a122-ad1c6984036c"></a>
### 6891-b8e7928/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:31259
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp ba75b6e273ff:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 2 minutes and 39 seconds
(2/13) Running aws-sdk-go tests ... done in 1 seconds
(3/13) Running aws-sdk-java tests ... done in 5 seconds
(4/13) Running aws-sdk-php tests ... done in 48 seconds
(5/13) Running aws-sdk-ruby tests ... done in 14 seconds
(6/13) Running mc tests ... done in 54 seconds
(7/13) Running minio-dotnet tests ... done in 5 minutes and 39 seconds
(8/13) Running minio-go tests ... done in 1 minutes and 49 seconds
(9/13) Running minio-java tests ... FAILED in 2 minutes and 52 seconds
{
  "name": "minio-java",
  "function": "presignedPostPolicy(PostPolicy policy)",
  "duration": 45,
  "status": "FAIL",
  "error": "java.net.SocketException: Connection reset >>> [java.net.SocketInputStream.read(SocketInputStream.java:210), java.net.SocketInputStream.read(SocketInputStream.java:141), okio.Okio$2.read(Okio.java:138), okio.AsyncTimeout$2.read(AsyncTimeout.java:236), okio.RealBufferedSource.indexOf(RealBufferedSource.java:344), okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:216), okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:210), okhttp3.internal.http1.Http1Codec.readResponseHeaders(Http1Codec.java:189), okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:75), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67), okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67), okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:120), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:92), okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:67), okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:185), okhttp3.RealCall.execute(RealCall.java:69), FunctionalTest.presignedPostPolicy_test(FunctionalTest.java:1900), FunctionalTest.runTests(FunctionalTest.java:2833), FunctionalTest.main(FunctionalTest.java:2942)]"
}

Executed 8 out of 13 tests successfully.
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :x: [more...](#2b4b09f3-8de3-4f51-98bc-9c3eb84fbed5) |
<a name="2b4b09f3-8de3-4f51-98bc-9c3eb84fbed5"></a>
### 6891-231d261/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30607
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 28ac24bc88d7:/mint/log /tmp/mint-logs'

(1/13) Running awscli tests ... done in 3 minutes and 36 seconds
(2/13) Running aws-sdk-go tests ... done in 1 seconds
(3/13) Running aws-sdk-java tests ... done in 7 seconds
(4/13) Running aws-sdk-php tests ... done in 42 seconds
(5/13) Running aws-sdk-ruby tests ... done in 3 seconds
(6/13) Running mc tests ... done in 23 seconds
(7/13) Running minio-dotnet tests ... done in 2 minutes and 40 seconds
(8/13) Running minio-go tests ... done in 53 seconds
(9/13) Running minio-java tests ... done in 2 minutes and 51 seconds
(10/13) Running minio-js tests ... FAILED in 31 seconds
{
  "name": "minio-js",
  "function": "listenBucketNotification(bucketName, prefix, suffix, events)",
  "args": "bucketName:minio-js-test-3b05b3ff-6e15-4fac-9990-f382933ef6e0, events: s3:ObjectCreated:*",
  "duration": 20028,
  "status": "FAIL",
  "error": "Error: notification not received at client.removeObject.err (test/functional-tests.js:1100:43) at ClientRequest.<anonymous> (node_modules/minio/dist/main/minio.js:419:9) at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:478:21) at HTTPParser.parserOnHeadersComplete (_http_common.js:99:23) at Socket.socketOnData (_http_client.js:367:20) at readableAddChunk (_stream_readable.js:176:18) at Socket.Readable.push (_stream_readable.js:134:10) at TCP.onread (net.js:559:20)"
}

Executed 9 out of 13 tests successfully.
```

**Comment by harshavardhana** [comment]
This PR is ready for review

**Comment by harshavardhana** [comment]
Used following compose for testing cc @poornas @nitisht 

```yml
version: '3'
services:

  minio-server-muc:
    image: custom-minio:latest
    # image: minio/minio
    ports:
      - 80:80
    volumes:
      - minio-server_muc_demo4_data_vol:/data
    entrypoint: >
        /bin/sh -c "
        curl https://raw.githubusercontent.com/nitisht/minio-compose/master/wait-for.sh -o wait-for.sh;
        chmod +x wait-for.sh;
        ./wait-for.sh etcd-server-muc:2379 -- /usr/bin/docker-entrypoint.sh minio gateway --address :80 nas /data/"
    environment:
      - "MINIO_ACCESS_KEY=minio"
      - "MINIO_SECRET_KEY=minio123"
      - "MINIO_ETCD_ENDPOINTS=http://etcd-server-muc:2379"
      - "MINIO_DOMAIN=tracecloud.test"
    networks:
      minio_network:
        ipv4_address: 172.18.0.2

  minio-server-sc:
    image: custom-minio:latest
    # image: minio/minio
    ports:
      - 90:80
    volumes:
      - minio-server_sc_demo4_data_vol:/data
    entrypoint: >
        /bin/sh -c "
        curl https://raw.githubusercontent.com/nitisht/minio-compose/master/wait-for.sh -o wait-for.sh;
        chmod +x wait-for.sh;
        ./wait-for.sh etcd-server-muc:2379 -- /usr/bin/docker-entrypoint.sh minio gateway --address :80 nas /data/"
    environment:
      - "MINIO_ACCESS_KEY=minio"
      - "MINIO_SECRET_KEY=minio123"
      - "MINIO_ETCD_ENDPOINTS=http://etcd-server-muc:2379"
      - "MINIO_DOMAIN=tracecloud.test"
    networks:
      minio_network:
        ipv4_address: 172.18.0.3

  etcd-server-muc:
    image: quay.io/coreos/etcd:v3.1.10
    ports:
      - 2379:2379
      - 2380:2380
    environment:
      - "ALLOW_NONE_AUTHENTICATION=yes"
      - "ETCD_ADVERTISE_CLIENT_URLS=http://etcd-server-muc:2379"
      - "ETCD_LISTEN_CLIENT_URLS=http://172.18.0.6:2379"
    networks:
      minio_network:
        ipv4_address: 172.18.0.6

volumes:
  minio-server_muc_demo4_data_vol:
  minio-server_sc_demo4_data_vol:

networks:
 minio_network:
   driver: bridge
   ipam:
    config:
      - subnet: 172.18.0.0/24
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

File: cmd/web-handlers.go [hunk 2]
```diff
@@ -1419,43 +1567,94 @@ func (web *webAPIHandlers) SetBucketPolicy(r *http.Request, args *SetBucketPolic
...
-	globalPolicySys.Set(args.BucketName, *bucketPolicy)
-	globalNotificationSys.SetBucketPolicy(ctx, args.BucketName, bucketPolicy)
+		// Parse validate and save bucket policy.
+		if err := objectAPI.SetBucketPolicy(ctx, args.BucketName, bucketPolicy); err != nil {
+			return toJSONError(err, args.BucketName)
+		}
+
+		globalPolicySys.Set(args.BucketName, *bucketPolicy)
+		globalNotificationSys.SetBucketPolicy(ctx, args.BucketName, bucketPolicy)
+	}
```
Comment (poornas) [inline cmd/web-handlers.go:1657]: @harshavardhana, PresignedGet also needs to be changed - if a share url is created on object belonging to cluster2 via cluster1 endpoint, anonymous download fails. It however succeeds if it belongs to same cluster

**Review by poornas — CHANGES_REQUESTED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

File: cmd/web-handlers.go [hunk 3]
```diff
@@ -99,11 +99,14 @@ func (web *webAPIHandlers) ServerInfo(r *http.Request, args *WebGenericArgs, rep
 	reply.MinioGlobalInfo = getGlobalInfo()
 	// If ENV creds are not set and incoming user is not owner
 	// disable changing credentials.
-	// TODO: fix this in future and allow changing user credentials.
 	v, ok := reply.MinioGlobalInfo["isEnvCreds"].(bool)
 	if ok && !v {
 		reply.MinioGlobalInfo["isEnvCreds"] = !owner
 	}
+	// if etcd is seet disallow changing credentials through UI
```
Comment (poornas) [inline cmd/web-handlers.go]: typo seet->set

Comment (harshavardhana) [reply cmd/web-handlers.go]: fixed

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by poornas — DISMISSED** [review]
Tested, LGTM

**Review by poornas — APPROVED** [review]

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-compression-xl.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-compression-fs.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Review by nitisht — APPROVED** [review]
LGTM & Tested

**Pull request merged**